### ISSUE TITLE
Add state=empty into file module

### ISSUE THREAD

**Issue by Nindaleth** [ansible/ansible-modules-core#902]
**Issue Type:**

Feature Idea

**Ansible Version:**

```
$ ansible --version
ansible 1.8.4
  configured module search path = None
```

**Environment:**

N/A

**Summary:**

In our playbooks we need to empty some folders, but currently our only option seems to be using `command` or `shell` module. It has its uses to leave the folder intact as there may have been some owner/permission changes, or alternatively it may be a mountpoint that is not even possible to delete.
Similar ideas might be had about files as well.
###### Proposed behaviour:
- Directories: remove all files and subdirectories inside so that the directory contains nothing
- Files: truncate the target file, if it was empty then touch it
- Symlinks/Hardlinks: check which of the previous 2 types it is, act accordingly

State change changed if previously non-empty, ok if previously empty, failed if still non-empty.
###### Example of usage:

```
- name: Clean folders
  file: path=/var/zookeeper state=empty

- name: Truncate logs
  file: path=/var/log/local6 state=empty
```

**Workaround (how to empty a directory completely):**

```
- find: path=/my/dir file_type=file
  register: files

- find: path=/my/dir file_type=directory
  register: directories

- file: state=absent path={{ item.path }}
  with_items: 
    - "{{ files }}"
    - "{{ directories }}"
```

Not sure if there is a reasonable workaround for emptying _multiple_ directories.

**Comment by pako-pl** [comment]
+1

**Comment by monkseal** [comment]
+1

**Comment by maticmeznar** [comment]
I just needed this myself. +1 indeed. Please make it happen!

**Comment by DavidGoodwin** [comment]
+1

**Comment by lerouxju** [comment]
+1

**Comment by gechr** [comment]
+1

**Comment by adeelshahid** [comment]
+1

**Comment by mutantcornholio** [comment]
+1

**Comment by Element20Three** [comment]
+1

**Comment by rafaelmagu** [comment]
Please click the Subscribe button on the right instead of adding `+1` as a comment.

**Comment by mutantcornholio** [comment]
Github truly needs issue voting...

**Comment by adeelshahid** [comment]
I need to hit the unsubscribe button with all these +1's coming into my inbox :smile:

**Comment by m11y** [comment]
+1

**Comment by tduchateau** [comment]
+1

**Comment by tersmitten** [comment]
+1

**Comment by thom8** [comment]
+1

**Comment by caljess599** [comment]
This would be great!

**Comment by bcoca** [comment]
```
- find: path=/my/dir
  register: files

- file: state=absent path={{item}}
  with_items: "{{files}}"
```

**Comment by Nindaleth** [comment]
Thanks Brian! That should help in most cases. With that in mind:

completely empty a directory
```
- find: path=/my/dir file_type=file
  register: files

- find: path=/my/dir file_type=directory
  register: directories

- file: state=absent path={{ item.path }}
  with_items: 
    - "{{ files }}"
    - "{{ directories }}"
```

Now, say I want to do the above for 10 directories, how could I go about that in a smart way? E.g. for an installed product clear any present logs, dumps, user configuration overrides, plugins, databases, downloads...

**Comment by thom8** [comment]
Something like this should work >

```
vars:
  empty_dirs:
    - "/dir1"
    - "/dir2"
    - ...

- file: state=absent path={{ item }}
  with_items: empty_dirs

- file: state=directory path={{ item }}
  with_items: empty_dirs
```

However this feature would be nice as it doesn't require multiple tasks or separate methods for files and directories.

**Comment by Nindaleth** [comment]
@thom8 That would discard the owner/group/access metadata, which is why I created this issue in the first place:

> It has its uses to leave the folder intact as there may have been some owner/permission changes, or alternatively it may be a mountpoint that is not even possible to delete.

**Comment by m-emelchenkov** [comment]
+1

**Comment by jc888** [comment]
+1

**Comment by st3fanK** [comment]
+1

**Comment by j-carpentier** [comment]
+1

**Comment by robputt** [comment]
+1

**Comment by caspark** [comment]
Until #3095 gets merged (fingers crossed), here's a workaround snippet which will empty a list of directories of files without deleting the directories themselves:

```
- name: Find files to purge
  find:
    path: "{{ item }}"
  register: find_results
  with_items: "{{ directories_to_purge_files_from }}"

  # ugly hack to extract paths; e.g. we want find_results.results[0].files[0].path (except for 0 to n in each case)
- name: Extract files to purge from find result
  local_action: command python -c "print '\n'.join([f['path'] for r in {{ find_results.results }} for f in r['files']])"
  register: find_results
  changed_when: false

- name: Ensure files to purge are absent
  file:
    state: absent
    path: "{{ item }}"
  with_items: "{{ find_results.stdout_lines|default([]) }}"

  # clean up just in case, since registered variables live on for the whole playbook
- name: Clean up temporary variable
  set_fact:
    find_results: ""
```

I want this as sort of a partial reset button, so I've got this as a role in `purge-files/tasks/main.yml` and it's set up in my playbook so that it requires a specific variable to be set:

```
- hosts: all
  roles:
  - role: purge-files
    tags: purge
    when: do_explicit_config_purge
    directories_to_purge_files_from:
      - /etc/nginx/sites-available
      - /etc/nginx/sites-enabled
      # you get the idea
```

So I can invoke it with something like `ansible-playbook myplaybook.yml --tags purge -e 'do_explicit_config_purge=true'`

(If you want to always run the "empty these directories" tasks, then you don't need `|default([])` - but I needed to add it to avoid a warning about `find_results.stdout_lines` being undefined when `not do_explicit_config_purge`; that smelt like a bug to me but seems to stem from `when` on roles being implemented as pushing the `when` down to tasks in a role, and `with_*` loops will be evaluated before `when`s.)

**Comment by Nindaleth** [comment]
@caspark Fingers crossed indeed, I'm waiting for at least a review of similar PR #933... for a year now :-)
EDIT: Maybe you could find some inspiration in that one? You handle the "absent" state which is great, maybe you could handle the "hard"/"link" state as well, as long as their target is eiter a file or a directory.

**Comment by Smana** [comment]
+1

**Comment by robputt** [comment]
@Nindaleth yeah I noticed that your PR had been marked as requiring rebase, I am not sure what the rules of engagement are for ansible repos, whether original contributor is required to rebase or if somebody who forms a release rebases and merges as required. Anyone know if there is there any documentation surrounding this for these projects?

**Comment by ssbarnea** [comment]
:+1:

**Comment by dylanmtaylor** [comment]
+1

**Comment by geoand** [comment]
+1

**Comment by michalkleiner** [comment]
+1

**Comment by elnur** [comment]
:+1:

**Comment by elnur** [comment]
Now all who +1'd this idea should go to #3095 and +1 it as well. :wink:

**Comment by richardasaurus** [comment]
+1

**Comment by jcranky** [comment]
+1

**Comment by Gerst20051** [comment]
👍

**Comment by byumov** [comment]
+1

**Comment by ddattee** [comment]
+1

**Comment by G-Goldstein** [comment]
+1

**Comment by sebastianfeldmann** [comment]
+1

**Comment by Fei-Guang** [comment]
+1

**Comment by hellonico** [comment]
+1

**Comment by dpesch** [comment]
+1

**Comment by ZeBidule** [comment]
+1

**Comment by gamename** [comment]
+1

**Comment by kartsims** [comment]
+1

**Comment by bachp** [comment]
+1

**Comment by hudecof** [comment]
+1

**Comment by m4dm4rtig4n** [comment]
+1

**Comment by Gerst20051** [comment]
why is this getting stalled?

**Comment by joequery** [comment]
+1

**Comment by pjmpsu** [comment]
+1

**Comment by Joyeric2** [comment]
Dearest one
My name is Joy Eric how are you doing today and how about your health,hope you are fine and good.Any way after going through your profile in this site i became interested in you.I will like to have you as my good friend whom i will like to share my life experience with.Your sex, age, race nor distance does not matter to me rather what matters most in a relationship is the maturity,truth and honest that exist between friends.So i will like you to contact me back through my private mail address(joyeric688@gma!l.com )So that i can tell you more about my self and send you more photos of me.Yours Sincerely,

**Comment by Joyeric2** [comment]
Dearest one
My name is Joy Eric how are you doing today and how about your health,hope you are fine and good.Any way after going through your profile in this site i became interested in you.I will like to have you as my good friend whom i will like to share my life experience with.Your sex, age, race nor distance does not matter to me rather what matters most in a relationship is the maturity,truth and honest that exist between friends.So i will like you to contact me back through my private mail address(joyeric688@gma!l.com )So that i can tell you more about my self and send you more photos of me.Yours Sincerely,

**Comment by pastushenko** [comment]
+1

**Comment by joaocenoura** [comment]
+1

**Comment by kamikaze** [comment]
+1

**Comment by jondubya** [comment]
+1

**Comment by lachlansimpson** [comment]
+1

**Comment by ckmcd** [comment]
Additional proposal.  We like to create "ballast" files.  The purpose of a ballast file is to allocate file system space that can easily by reclaimed by deleting the file.  The proposal is to add a size option that would truncate the file to a specified size, extending the current allocated size if necessary.  Of course this only makes sense for files not directories.

**Comment by radzhome** [comment]
+1

**Comment by erikdw** [comment]
@ckmcd : you should probably file a separate issue for that.  Also, everyone putting in "+1" isn't really helpful, it's spamming the content of the issue. There's already a "reaction" thing at the bottom of the initial description.  I would just 👍 in there...

**Comment by azizfcb** [comment]
+1

**Comment by grantlittle** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by danidano** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by richardasaurus** [comment]
+9000

**Comment by hudecof** [comment]
Hi,

I wrote support for state=empty based on the https://github.com/ansible/ansible-modules-core/pull/3095

https://github.com/hudecof/ansible-modules-core/commit/7bd11b6d70ec44bb858020525ce93e38482e02b2

But tehre are following questions:
- what to do with symlinks ? Should I follow them and empty the linked file/dir
- what to do with hardlinks ? Should I follow them and empty the linked file

**Comment by erikdw** [comment]
@hudecof : I think the main problem you will face is getting someone from the ansible core team to review your change.  Since @robputt796 seems to have abandoned #3095, you @hudecof should probably create a new PR and explain the difference between #3095 & what you've done.

As for the 2 questions you posed: I would have the module behave the same as `rm -r`.  So you definitely wouldn't delete files pointed to by symbolic links.  As for hardlinks, I'm not sure what it would even _mean_ to "follow" it.  i.e., a hardlink is just a pointer to the same physical file as the original one (the hardlink is created by having the new file's inode value set to the original file's inode value).  There is no cookie crumb left to easily find the original file, and it would be wrong to go and delete that anyways.

**Comment by hudecof** [comment]
@erikdw 
for HARDLINK, only file version, I could add the code to truncate the file
for SYMLINK it could be more difficult to follow it. The ansible has some code to do that, I need to investigate how it really works.

And about the code review. I agree, few years ago, the ansible  was small project, it was quite easy to contribute. 

So I will try to do **rm -rf /path/*** on symlink to directory and see what happend ;)

**Comment by erikdw** [comment]
@hudecof:

>  for HARDLINK, only file version, I could add the code to truncate the file

I wouldn't do that.  Per the comments at the top of your code, this is just supposed to recursively remove files from the directory.  Remove them, not truncate them.  So when you remove a hardlink from the directory, it wouldn't affect the other places that hold reference to that physical file.

> So I will try to do rm -rf /path/\* on symlink to directory and see what happend ;)

`rm -rf` does not follow symlinks, it just removes the link from the dir it is in.  It would be really annoying if it did that.

**Comment by hudecof** [comment]
@erikdw 
I know how it works, but my question was a little bit different. Im not talking about the direcotry content itselt, but the PATH argument.

So what to do, if you try to empty/truncate link itself.
The @robputt796  code ignores this cases. Example

```
ansible localhost -m file -a "state=empty path=/tmp/some.symlink"
ansible localhost -m file -a "state=empty path=/tmp/some.hardlink"
```

The **path** argument  could be  only real **directory** or **file**, at leat at this moment.

**Comment by Nindaleth** [comment]
@hudecof #3095 seems abandoned and I'm about to close/abandon my own #933, please go forward with your PR whenever you feel ready

**Comment by hudecof** [comment]
@Nindaleth I will check also your code and take the best.

But it seems to be long run task from the ansible team to merge the code ;(

**Comment by erikdw** [comment]
> my question was a little bit different. Im not talking about the direcotry content itselt, but the PATH argument.

@hudecof : ahhhh, that was not clear to me, sorry for the misunderstanding. 

So I now see why you are unsure about how to handle it.  Not obvious what to do for these suckers (deleting the file vs. emptying the pointed to thing).

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dentarg** [comment]
about the workaround given in https://github.com/ansible/ansible-modules-core/issues/902#issuecomment-164043327, with ansible 2.1.2.0 it should be:

remove files from dir

```
- find: path=/my/dir
  register: files

- file: state=absent path={{item.path}}
  with_items: "{{files.files}}"
```

**Comment by lukasojd** [comment]
+1

**Comment by valentin-nasta** [comment]
+1

**Comment by kkarimi** [comment]
+1

**Comment by kkarimi** [comment]
+1

**Comment by kkarimi** [comment]
+1

**Comment by vanhuto** [comment]
+1

**Comment by lizhongz** [comment]
+1

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by erikdw** [comment]
This issue has been moved to the ansible repo: https://github.com/ansible/ansible/issues/18910

**Comment by erikdw** [comment]
Unfortunately some comments got lost along the way since the mover failed with an error before moving all comments:
![screenshot 2016-12-07 11 17 49](https://cloud.githubusercontent.com/assets/1111441/20982847/03359640-bc6f-11e6-8216-e0337503dfd1.png)

**Comment by valugi** [comment]
+1

**Comment by Imkerix** [comment]
+1

**Comment by Harosgr** [comment]
Do we have any progress with that feature? Looking at https://github.com/ansible/ansible-modules-core/pull/5652 it seems that the core team isn't interested at the moment. I understand that the workaround with find works, but it is... a workaround. Why should someone use two extra tasks when the job could be done in one?

**Comment by mijohnst** [comment]
+1

**Comment by aaratn** [comment]
+1

**Comment by copacetic** [comment]
+1

**Comment by florres** [comment]
+1

**Comment by Fobhep** [comment]
+1

**Comment by raasss** [comment]
I was looking for similar solution and for me synchronize module was enough. Made empty directory on ansible host and synchronized it to remote dir. For example:

- name: Empty remote directory
  synchronize:
    src: empty/
    dest: "{{ remote_dir }}/"
    delete: yes
    recursive: yes

**Comment by hudecof** [comment]
@raasss could be, thanks

**Comment by dagwieers** [comment]
This issue has been moved to the ansible repo: ansible/ansible#18910
